doc/services/storage/zms/zms.rst

```diff
@@ -0,0 +1,432 @@
+===============
+
+To close a sector a close ATE is added at the end of the sector and it must have the same cycle
+counter as the empty ATE.
+When closing a sector, all the remaining space that has not been used is filled with garbage data
+to avoid having old ATEs with a valid cycle counter.
+
+Triggering Garbage collection
+=============================
+
+Some applications need to make sure that storage writes have a maximum defined latency.
+When calling a ZMS write, the current sector could be almost full and we need to trigger the GC
+to switch to the next sector.
+This operation is time consuming and it will cause some applications to not meet their real time
+constraints.
+ZMS adds an API for the application to get the current remaining free space in a sector.
+The application could then decide when needed to switch to the next sector if the current one is
+almost full and of course it will trigger the garbage collection on the next sector.
+This will guarantee the application that the next write won't trigger the garbage collection.
+
+ATE (Allocation Table Entry) structure
+======================================
+
+An entry has 16 bytes divided between these variables :
+
+.. code-block:: c
+
+	struct zms_ate {
+		uint8_t crc8;      /* crc8 check of the entry */
+		uint8_t cycle_cnt; /* cycle counter for non erasable devices */
+		uint32_t id;       /* data id */
+		uint16_t len;      /* data len within sector */
+		union {
+			uint8_t data[8]; /* used to store small size data */
+			struct {
+				uint32_t offset; /* data offset within sector */
+				union {
+					uint32_t data_crc; /* crc for data */
+					uint32_t metadata; /* Used to store metadata information
+							    * such as storage version.
+							    */
+				};
+			};
+		};
+	} __packed;
+
+.. note:: The data CRC is checked only when the whole data of the element is read.
+   The data CRC is not checked for a partial read, as it is computed for the complete set of data.
+
+.. note:: Enabling the data CRC feature on a previously existing ZMS content without
+   data CRC will make all existing data invalid.
+
+.. _free-space:
+
+Available space for user data (key-value pairs)
+***********************************************
+
+For both scenarios ZMS should have always an empty sector to be able to perform the garbage
+collection.
+So if we suppose that 4 sectors exist in a partition, ZMS will only use 3 sectors to store
+Key-value pairs and keep always one (rotating sector) empty to be able to launch GC.
+
+.. note:: The maximum single data length that could be written at once in a sector is 64K
+   (This could change in future versions of ZMS)
+
+Small data values
+=================
+
+For small data values (<= 8 bytes), the data is stored within the entry (ATE) itself and no data
+is written at the top of the sector.
+ZMS has an entry size of 16 bytes which means that the maximum available space in a partition to
+store data is computed in this scenario as :
```

Comment:
No the calculation is true for all values which size is <= 8 bytes

Reply:
Well I precised in the formula that this is the maximum available space for data.
But of course it will depend on the size of the data.
I will change the formula to make it more clear